### PR TITLE
Add performance.mark and performance.measure for performance browser tool integration #617

### PR DESCRIPTION
**Pull request by MSNev**

### CODE REVIEW

File: extensions/applicationinsights-perfmarkmeasure-js/README.md [hunk 1]
```diff
@@ -0,0 +1,84 @@
...
+
+[![npm version](https://badge.fury.io/js/%40microsoft%2Fapplicationinsights-perfmarkmeasure-js.svg)](https://badge.fury.io/js/%40microsoft%2Fapplicationinsights-perfmarkmeasure-js)
+
+Performance Mark and Measure Manager Plugin for the Application Insights Javascript SDK is a replacement for the default Performance manager to 
+use the ```window.performance```, ```mark``` and ```measure``` api calls for the browsers.
+
+It is intended that this plugin is to be used during development/testing only as enabling performance monitoring can affect the performance of the system.
+And there is a limit on the number of performance events that browsers will retain.
+
+## Getting Started
```
Comment (MSNev) [inline extensions/applicationinsights-perfmarkmeasure-js/README.md:11]: Still needs some more documentation here

File: extensions/applicationinsights-perfmarkmeasure-js/Tests/Unit/src/MarkMeasureTests.ts [hunk 2]
```diff
@@ -0,0 +1,371 @@
...
+import { ITelemetryItem, AppInsightsCore, IPlugin, IConfiguration, DiagnosticLogger, IAppInsightsCore, ITelemetryPluginChain} from '@microsoft/applicationinsights-core-js';
+import { PerfMarkMeasureManager } from '../../../src/PerfMarkMeasureManager';
+
+export interface PerfMeasures {
+    name: string;
+    from: string;
+    to: string;
+};
+
+export class MarkMeasureTests extends AITestClass {
```
Comment (MSNev) [inline extensions/applicationinsights-perfmarkmeasure-js/Tests/Unit/src/MarkMeasureTests.ts:15]: Still want to add a few more tests using doPerf() directly

File: extensions/applicationinsights-perfmarkmeasure-js/Tests/Unit/src/MarkMeasureTests.ts [hunk 3]
```diff
@@ -0,0 +1,956 @@
...
+                        numEnd++;
+                    }
+                }
+
+                Assert.equal(numStart, numEnd, "Should be same number of start and end marks");
+                Assert.equal(true, this._measures.length > 0);
+
+                for (let lp = 0; lp < this._measures.length; lp++) {
+                    let measure = this._measures[lp];
+                    // Making sure there are no "end" marks
```
Comment (xiao-lix) [inline extensions/applicationinsights-perfmarkmeasure-js/Tests/Unit/src/MarkMeasureTests.ts:316]: nit: // Making sure there are "end" marks

File: extensions/applicationinsights-perfmarkmeasure-js/README.md [hunk 4]
```diff
@@ -0,0 +1,164 @@
...
+
+## Configuration Options
+
+All of the configuration options are optional
+
+### IPerfMarkMeasureConfiguration
+
+| Name | Type | Description
+|------|-----------|--------------------
+| useMarks | boolean | Should the Performance manager create and user window.performance.mark(), defaults to true
```
Comment (xiao-lix) [inline extensions/applicationinsights-perfmarkmeasure-js/README.md:94]: nit : create and **use**

File: extensions/applicationinsights-perfmarkmeasure-js/README.md [hunk 5]
```diff
@@ -0,0 +1,164 @@
...
+| markPrefix | string | Identifies the prefix for the mark, defaults to "ai.prfmrk.", the event name is appended for the mark
+| uniqueNames | boolean | Make the marks and measures unique by appending a numeric value to the prefix value, defaults to false. Marks and measure for the same perfEvent will be assigned the same unique numeric value
+| markNameMap | { [key: string]: string } | Provides a mapping between the internal perf names and the value used to create the mark, when a map is provided but no mapping is present that event will be ignored.
+| useEndMarks | boolean | Should the Performance manager create a mark when the event is fired, defaults to false
+| markEndPrefix | string | Identifies the prefix for the "end" mark of a perf event, defaults to "ai.prfmrk.end.", the event name is appended for the mark
+| useMeasures | boolean | Should the Performance manager create and user window.performance.measure(), defaults to true
```
Comment (xiao-lix) [inline extensions/applicationinsights-perfmarkmeasure-js/README.md:100]: nit: use

File: extensions/applicationinsights-perfmarkmeasure-js/Tests/Unit/src/MarkMeasureTests.ts [hunk 6]
```diff
@@ -0,0 +1,956 @@
...
+                        numEnd++;
+                    }
+                }
+
+                Assert.equal(numStart, numEnd, "Should be same number of start and end marks");
+
+                Assert.equal(true, this._measures.length > 0);
+                for (let lp = 0; lp < this._measures.length; lp++) {
+                    let measure = this._measures[lp];
+                    // Making sure there are no "end" marks
```
Comment (xiao-lix) [inline extensions/applicationinsights-perfmarkmeasure-js/Tests/Unit/src/MarkMeasureTests.ts:249]: nit : there are "end" marks

File: extensions/applicationinsights-perfmarkmeasure-js/Tests/Unit/src/MarkMeasureTests.ts [hunk 7]
```diff
@@ -0,0 +1,956 @@
...
+            test: () => {
+                const manager = new PerfMarkMeasureManager({
+                    useMarks: true,
+                    useEndMarks: true,
+                    uniqueNames: true
+                });
+
+                Assert.equal(0, this._marks.length);
+                doPerf(manager, () => "test", (perfEvent) => {
+                    Assert.equal(1, this._marks.length > 0);
```
Comment (xiao-lix) [inline extensions/applicationinsights-perfmarkmeasure-js/Tests/Unit/src/MarkMeasureTests.ts:426]: Should the assert be `Assert.equal(1, this._marks.length);`?

File: extensions/applicationinsights-perfmarkmeasure-js/Tests/Unit/src/MarkMeasureTests.ts [hunk 8]
```diff
@@ -0,0 +1,956 @@
...
+        this.testCase({
+            name: "Test with direct doPerf(): Disable Measures",
+            test: () => {
+                const manager = new PerfMarkMeasureManager({
+                    useMeasures: false
+                });
+
+                Assert.equal(0, this._marks.length);
+                doPerf(manager, () => "test3", (perfEvent) => {
+                    Assert.equal(1, this._marks.length > 0);
```
Comment (xiao-lix) [inline extensions/applicationinsights-perfmarkmeasure-js/Tests/Unit/src/MarkMeasureTests.ts:506]: `Assert.equal(1, this._marks.length);`?

File: extensions/applicationinsights-perfmarkmeasure-js/Tests/Unit/src/MarkMeasureTests.ts [hunk 9]
```diff
@@ -0,0 +1,956 @@
...
+                    uniqueNames: true,
+                    markNameMap: {
+                        "test": "mapped1",
+                        "test3": "mapped3"
+                    }
+                });
+
+                Assert.equal(0, this._marks.length);
+                doPerf(manager, () => "test", (perfEvent) => {
+                    Assert.equal(1, this._marks.length > 0);
```
Comment (xiao-lix) [inline extensions/applicationinsights-perfmarkmeasure-js/Tests/Unit/src/MarkMeasureTests.ts:598]: Assert.equal(1, this._marks.length);

File: extensions/applicationinsights-perfmarkmeasure-js/Tests/Unit/src/MarkMeasureTests.ts [hunk 10]
```diff
@@ -0,0 +1,956 @@
...
+                    },
+                    measureNameMap: {
+                        "test": "measure1",
+                        "test2": "measure2"
+                    }
+                });
+
+                Assert.equal(0, this._marks.length);
+                doPerf(manager, () => "test", (perfEvent) => {
+                    Assert.equal(1, this._marks.length > 0);
```
Comment (xiao-lix) [inline extensions/applicationinsights-perfmarkmeasure-js/Tests/Unit/src/MarkMeasureTests.ts:646]: same here

File: extensions/applicationinsights-perfmarkmeasure-js/Tests/Unit/src/MarkMeasureTests.ts [hunk 11]
```diff
@@ -0,0 +1,956 @@
...
```
Comment (xiao-lix) [inline extensions/applicationinsights-perfmarkmeasure-js/Tests/Unit/src/MarkMeasureTests.ts:698]: same here

File: extensions/applicationinsights-perfmarkmeasure-js/src/interfaces/IPerfMarkMeasureConfiguration.ts [hunk 12]
```diff
@@ -0,0 +1,59 @@
...
+ * @copyright Microsoft 2021
+ */
+
+ /**
+  * PerfMarkMeasure Configuration
+  */
+ export interface IPerfMarkMeasureConfiguration {
+
+    /**
+     * Should the Performance manager create and user window.performance.mark(), defaults to true
```
Comment (xiao-lix) [inline extensions/applicationinsights-perfmarkmeasure-js/src/interfaces/IPerfMarkMeasureConfiguration.ts:11]: nit: use

**Comment by xiao-lix** [comment]
Read through the PR, can't find anything other than nit stuff 😄 nice feature. 👍

**Review by xiao-lix — APPROVED** [review]

File: extensions/applicationinsights-perfmarkmeasure-js/Tests/Unit/src/MarkMeasureTests.ts [hunk 13]
```diff
@@ -0,0 +1,956 @@
...
+            name: "Test with direct doPerf(): Specific enable start marks with unique",
+            test: () => {
+                const manager = new PerfMarkMeasureManager({
+                    useMarks: true,
+                    uniqueNames: true
+                });
+
+                Assert.equal(0, this._marks.length);
+                doPerf(manager, () => "test", (perfEvent) => {
+                    Assert.equal(1, this._marks.length > 0);
```
Comment (xiao-lix) [inline extensions/applicationinsights-perfmarkmeasure-js/Tests/Unit/src/MarkMeasureTests.ts:372]: same here

**Pull request merged**